### PR TITLE
[WIP] Add firewall service

### LINKED ISSUE #422: Firewall: deploy port rules for every component
A lot of companies enforce strong security requirements, including strict control and documentation of firewall rules. Offering a deployment of firewall port rules would ease TDP integration.

### PR DESCRIPTION
**Pull request by Nuttymoon**
#### Which issue(s) this PR fixes:

Fixes #422 

#### Additional comments:

**This PR is a follow up on PR #452**

**Work in progress:** the firewall rules are only implemented for HDFS DataNode. The goal of this PR for now is to get feedback on the implementation.

The implementation proposed here inherits the concepts of [firewalld](https://firewalld.org/): [zones](https://firewalld.org/documentation/zone/) and [services](https://firewalld.org/documentation/service/). A service corresponds to a list of ports (+protocols) and a zone corresponds to a list of sources (~= IPs) and the rules that apply to them when accessing services on the host.

##### Possible negative impacts

Deploying erroneous firewall rules on hosts can have disastrous consequences, notably forbidding node operators to connect via SSH. I suggest that we implement a check that determines if Ansible will still be able to connect after the firewall rules deployment and if not abort it (see #480).

##### `firewall` role

A new `firewall` role is added to TDP. This role will always have to be run prior to services deployment. The role has been written to support multiple firewall services as a backend. For now only `firewalld` is supported. One can change which firewall service is used with `firewall_service` and the routing logic is managed in `firewall.generic`.

```
roles/firewall
├── firewalld
│   ├── handlers
│   ├── tasks
│   └── templates
└── generic
    └── tasks
```

##### Cluster level firewall

`firewall_tdp_managed` controls if TDP should generate firewall rules on the cluster nodes:
- If `yes`, TDP ensures that all rules defined in `firewall_rules` are created on the right hosts and that `firewalld` is started.
- If `no`, TDP ensures that all previously generated firewall rules are deleted.

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L5-L10

##### Firewall zones

By default, TDP will create 2 zones:

- `public`: This zone corresponds to all the hosts outside of the TDP cluster that need access to its client endpoints (e.g. HDFS Web UI, Hive JDBC/ODBC).
- `internal`: This zone corresponds to all the TDP cluster hosts. Internal ports need to be opened inside the cluster.

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L16-L26

Each zone has 4 properties:

- `name`: A name for the zone
- `short`: A short description of the rule that will be used in the `firewalld` zone definition
- `groups`: The Ansible groups that are part of this zone. TDP will take care of mapping these groups to [firewalld ipsets](https://firewalld.org/documentation/ipset/).
- `addresses`: The other addresses that are part of this zone. Can be in the format `address` or `address/mask`.

##### Firewall rules

All open port rules are listed in the `firewall_rules` variable, with one list for each component. E.g. for `hdfs_dn` (HDFS DataNode):

https://github.com/TOSIT-IO/tdp-collection/blob/14809dd1a1c8c91778bf5ebe99b60a7daa0649fb/tdp_vars_defaults/firewall/firewall.yml#L40-L48

Each rule has 5 properties:

- `short`: A short description of the rule that will be used in the `firewalld` service definition
- `port`: The port to open and secure
- `protocol`: The protocol authorized on this port
- `zones`: The zones from which the port will be accessible.
- `state`: `present` or `absent`. If `absent`, TDP will make sure that the port is removed from all rules.

##### Trying this out

If you want to try this on a TDP cluster, run `playbooks/meta/firewall.yml`. This will deploy firewall rules on `hdfs_dn` hosts and start `firewalld`. Note that other services installed on the same hosts (e.g. YARN RM) will stop functioning as no firewall rules will exist for them.

##### Low level implementation

TDP uses `firewalld` ipsets, services, and zones to set firewall rules (all generated files are prefixed with `firewalld_files_prefix`):

- **Firewalld ipsets:** each Ansible group that is used in `firewall_zones` will be mapped to an ipset in `/etc/firewalld/ipsets`. E.g. `/etc/firewalld/ipsets/tdp_tdp_cluster.xml`:

  ```xml
  <?xml version="1.0" encoding="utf-8"?>
  <ipset type="hash:net">
    <short>tdp_cluster Ansible group</short>
    <entry>192.168.56.13</entry>
    ...
    <entry>192.168.56.16</entry>
  </ipset>
  ```

- **Firewalld services:** each TDP component has 1 service (in `/etc/firewalld/services`) for each of its `firewalld_rules`. A service contains the port number and protocol. E.g. `/etc/firewalld/services/tdp_hdfs_dn_data_port.xml`:

  ```xml
  <?xml version="1.0" encoding="utf-8"?>
  <service>
    <short>HDFS DataNode data transfer port</short>
    <port protocol="tcp" port="9866" />
  </service>
  ```

- **Firewalld zones:** Each zone in `firewall_zones` has 1 zone in `/etc/firewalld/zones` that defines the sources allowed to the zone's services. E.g. `/etc/firewalld/zones/tdp_internal.xml`:

  ```xml
  <?xml version='1.0' encoding='UTF-8'?>
  <zone>
    <short>TDP cluster internal zone</short>
    <source ipset="tdp_tdp_cluster"/>
    <service name="tdp_hdfs_dn_data_port"/>
    <service name="tdp_hdfs_dn_ipc_port"/>
    ...
  </zone>
  ```

#### License agreement

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Comment by Nuttymoon** [comment]
I just noticed that the role does not work in Ansible 2.9 because https://github.com/ansible/ansible/pull/64959 has been fixed in Ansible 2.10.

I am looking for a workaround.

**Comment by Nuttymoon** [comment]
> I just noticed that the role does not work in Ansible 2.9 because [ansible/ansible#64959](https://github.com/ansible/ansible/pull/64959) has been fixed in Ansible 2.10.
> 
> I am looking for a workaround.

- [x] Done

**Comment by Nuttymoon** [comment]
To do:
- [ ] Remove firewall rules using the entire file names instead of `{{ firewalld_files_prefix }}*.xml` to avoid deleting files generated by the client

**Comment by Nuttymoon** [comment]
Rebased on master

File: roles/firewall/firewalld/tasks/config.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        services: |
+          {{ host_firewalld_rules | json_query("[?contains(zones, '" + item.name + "')]") }}
+      loop: "{{ firewalld_zones }}"
+      loop_control:
+        label: "{{ firewalld_files_prefix }}{{ item.name }}"
+      notify: "Reload firewalld"
+  when: firewall_tdp_managed
+
+- block:
+    - name: Delete all TDP firewalld ipsets
```
Comment (rpignolet) [inline roles/firewall/firewalld/tasks/config.yml:56]: `firewalld` should be reloaded if you remove TDP firewalld files.

**Review by rpignolet — CHANGES_REQUESTED** [review]

File: roles/firewall/firewalld/tasks/config.yml (hunk 1 above)
Comment (Nuttymoon) [reply roles/firewall/firewalld/tasks/config.yml:56]: I think that we should let the user handle the reload of firewalld. Reloading/restarting firewalld removes all existing non-persisted rules.

Comment (rpignolet) [reply roles/firewall/firewalld/tasks/config.yml:56]: This case is when you remove TDP firewall config file, if you use the `notify` attribute, it reload the firewall only if a TDP file is removed. What is "existing non-persisted rules" ? When the firewall is manage by TDP you handle the non-persisted rules when you reload ?

Comment (Nuttymoon) [reply roles/firewall/firewalld/tasks/config.yml:56]: They are rules that have been set at runtime in firewalld. When the firewall is managed by TDP the user can expect that we will make changes to it and reload the service (we should also document this behaviour).

I think that when TDP firewall management is disabled, the user will expect that it doesn't interact with the service.

Comment (Nuttymoon) [reply roles/firewall/firewalld/tasks/config.yml:56]: We could move the TDP files removing logic to a dedicated playbook that the user can run when needed.

Comment (rpignolet) [reply roles/firewall/firewalld/tasks/config.yml:56]: IMO if we keep the remove logic we should reload the firewall service, if not the firewall will not be consistent. If we do not want to reload the service, we should not remove firewall files.

Comment (Nuttymoon) [reply roles/firewall/firewalld/tasks/config.yml:56]: As discussed today, I am gonna remove the logic from the playbooks and keep the tasks for further documentation.

Comment (Nuttymoon) [reply roles/firewall/firewalld/tasks/config.yml]: The clean-up tasks have been moved to `clean-up` playbooks

**Comment by Nuttymoon** [comment]
As discussed previously, I will create several PRs for this branch to add firewall rules in small batches.

**Pull request closed without merging**